sendinvoiceless/sendinvoiceless.py

```diff
@@ -0,0 +1,119 @@
+                fee = Millisatoshi(ch['base_fee_millisatoshi'])
+                fee += msatoshi * ch['fee_per_millionth'] // 1000000
+                if ch['source'] == payload['nodeid']:
+                    if fee <= payload['msatoshi']:
+                        fee = payload['msatoshi']
+                    else:
+                        raise RpcError("sendinvoiceless", payload, {'message': 'Insufficient sending amount'})
+                msatoshi += fee
+                delay += ch['delay']
+                r['direction'] = int(ch['channel_flags']) % 2
+
+
+def find_worst_channel(route, nodeid):
+    worst = None
+    worst_val = Millisatoshi(0)
+    for i in range(1, len(route)):
+        if route[i - 1]['id'] == nodeid:
+            continue
+        val = route[i - 1]['msatoshi'] - route[i]['msatoshi']
+        if val > worst_val:
+            worst = route[i]['channel'] + '/' + str(route[i]['direction'])
+            worst_val = val
+    return worst
+
+
+def sendinvoiceless_fail(plugin, label, payload, success_msg, error=None):
+    try:
+        plugin.rpc.delinvoice(label, 'unpaid')
+    except RpcError as e:
+        # race condition: waitsendpay timed out, but invoice get paid
+        if 'status is paid' in e.error.get('message', ""):
+            return success_msg
+    if error is None:
+        error = RpcError("sendinvoiceless", payload, {'message': 'Sending failed'})
+    raise error
+
+
+@plugin.method("sendinvoiceless")
+def sendinvoiceless(plugin, nodeid, msatoshi: Millisatoshi, maxfeepercent="0.5",
+                    retry_for=60, exemptfee: Millisatoshi=Millisatoshi(5000)):
+    """Invoiceless payment with circular routes.
+
+    This tool sends some msatoshis without needing to have an invoice from the receiving node.
+
+    """
+    payload = {
+        "nodeid": nodeid,
+        "msatoshi": msatoshi,
+        "maxfeepercent": maxfeepercent,
+        "retry_for": retry_for,
+        "exemptfee": exemptfee
+    }
+    myid = plugin.rpc.getinfo().get('id')
+    label = "InvoicelessChange-" + str(uuid.uuid4())
+    description = "Sending %s to %s" % (msatoshi, nodeid)
+    change = Millisatoshi(1000)
+    invoice = plugin.rpc.invoice(change, label, description, int(retry_for) + 60)
+    payment_hash = invoice['payment_hash']
+    plugin.log("Invoice payment_hash: %s" % payment_hash)
+    success_msg = ""
+    try:
+        excludes = []
+        start_ts = int(time.time())
+        while int(time.time()) - start_ts < int(retry_for):
+            forth = plugin.rpc.getroute(nodeid, msatoshi + change, riskfactor=10, exclude=excludes)
+            back = plugin.rpc.getroute(myid, change, riskfactor=10, fromid=nodeid, exclude=excludes)
+            route = forth['route'] + back['route']
+            setup_routing_fees(plugin, route, change, payload)
+            fees = route[0]['msatoshi'] - route[-1]['msatoshi'] - msatoshi
+            if fees > exemptfee and fees > msatoshi * float(maxfeepercent) / 100:
```

Comment:
This will (currently) fail due to unresolved https://github.com/ElementsProject/lightning/pull/2601

I suggest the following workaround (as included in my improve rebalance PR):
```Python
if fees > exemptfee and int(fees) > int(msatoshi) * float(maxfeepercent) / 100:
```

Reply:
Nice catch! I'm already waiting for the next pylightning release. Currently it causes exception if someone gives the rebalance/sendinvoiceless `msatoshi` in the form of "1sat". https://github.com/ElementsProject/lightning/pull/2581